DOCSP-15907 node aggregation pg (#212)
mongodb/docs-node: source/fundamentals/aggregation.txt

- rename fields
- calculate fields
- summarize data
- group values

Aggregation operations have some :manual:`limitations </core/aggregation-pipeline-limits/>`:

- Returned documents must not violate the :manual:`BSON-document size limit </reference/limits/#mongodb-limit-BSON-Document-Size>`
  of 16 megabytes.

- Pipeline stages have a memory limit of 100 megabytes by default. If necessary, you may exceed this limit by setting the ``allowDiskUse``
  property of ``AggregateOptions`` to ``true``. See the
  :node-api-4.0:`AggregateOptions API documentation <interfaces/aggregateoptions.html>`
  for more details.

.. important:: ``$graphLookup`` exception

   The :manual:`$graphLookup
   </reference/operator/aggregation/graphLookup/>` stage has a strict
   memory limit of 100 megabytes and will ignore ``allowDiskUse``.

Useful References
~~~~~~~~~~~~~~~~~

- :manual:`Expression operators </reference/operator/aggregation/>`
- :manual:`Aggregation pipeline </core/aggregation-pipeline/>`
- :manual:`Aggregation stages </meta/aggregation-quick-reference/#stages>`
- :manual:`Operator expressions </meta/aggregation-quick-reference/#operator-expressions>`

Runnable Examples
-----------------

> The following example inserts sample data into the ``restaurants``
> collection of the ``aggregation`` database.

Reviewer: ```suggestion
The examples use a fictional sample data set about restaurants. The following code inserts the sample data into the ``restaurants``
collection of the ``aggregation`` database.
```

I think this makes it sound like the following code snippet demonstrates aggregation. I think this should be clarified so it is clear that this is only here to create a sample dataset so that aggregation can be demonstrated later.

Author: I separated the sentences so that the insertion sentence comes right before the code.

.. tip::

>    For information on connecting to your MongoDB deployment, see the :doc:`Connection Guide </fundamentals/connection>`.

Reviewer: ```suggestion
   For more information on connecting to your MongoDB deployment, see the :doc:`Connection Guide </fundamentals/connection>`.
```

Nit: For whatever reason, "for information" seems off to me. It makes me think that no information about connecting to MongoDB is contained on this page, which I don't think is true (specifying DB and collection is shown).

Author: accepted

===========
Aggregation
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use **aggregation operations** in
the MongoDB Node.js driver.

Aggregation operations are expressions you can use to produce summarized

> and computed results in MongoDB. MongoDB's aggregation framework

Author: accepted

allows you to create a pipeline that consists of one or more stages,
each of which performs a specific operation on your data.

You can think of the aggregation framework as similar to an automobile factory.
Automobile manufacturing requires the use of assembly stations organized

> into assembly lines. Each station has specialized tools, such as
> drills, welders, or other machinery. The factory transforms and

Reviewer: ```suggestion
into assembly lines. Each station has specialized tools to do a specific job, such as
drills and welders. The factory transforms and
```

I think "other machinery" should be removed. I think "machinery" is a different category than "tool", and I think "other machinery" is vague.

Author: ok

assembles the initial parts and materials into finished products.

The **aggregation pipeline** is the assembly line, **aggregation stages** are the assembly stations, and
**operator expressions** are the specialized tools.

Aggregation vs. Query Operations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Using query operations, such as the ``find()`` method, you can:

- select *which documents* to return
- select *which fields* to return
- sort the results

Using ``aggregation`` operations, you can:

> - perform all ``find`` operations

Author: accepted

  for more details.

.. important:: ``$graphLookup`` exception

   The :manual:`$graphLookup
   </reference/operator/aggregation/graphLookup/>` stage has a strict
   memory limit of 100 megabytes and will ignore ``allowDiskUse``.

Useful References
~~~~~~~~~~~~~~~~~

- :manual:`Expression operators </reference/operator/aggregation/>`
- :manual:`Aggregation pipeline </core/aggregation-pipeline/>`
- :manual:`Aggregation stages </meta/aggregation-quick-reference/#stages>`
- :manual:`Operator expressions </meta/aggregation-quick-reference/#operator-expressions>`

Runnable Examples
-----------------

Create a new file and connect to your MongoDB deployment to follow this
example.

.. tip::

   For information on connecting to your MongoDB deployment, see the :ref:`Connection Guide <connection>`.
   
Insert Sample Data
~~~~~~~~~~~~~~~~~~

The example uses sample data about restaurants. The following code
inserts the data into the ``restaurants`` collection of the ``aggregation`` database:

.. literalinclude:: /code-snippets/aggregation/agg.js
   :start-after: begin data insertion
   :end-before: end data insertion
   :language: javascript
   :dedent:

> 

Author: I'm in favor of leaving the tip there linking to the connection guide, do you think that the tip could just stay there even without that text?

===========
Aggregation

> ===========

Author: yes, I am planning on adding it to the TOC and fundamentals landing page when #211 is merged

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use **aggregation operations** in
the MongoDB Node.js driver.

Aggregation operations are expressions that you can execute against data in your MongoDB

> collections to return summarized and computed results. MongoDB's aggregation framework

Author: I changed the wording, but I think your suggestion might imply that aggregations alter data in the collections.

> collections to return summarized and computed results. MongoDB's aggregation framework

Reviewer: I am not sure "produce ... results in MongoDB" removes the implication. Are the results produced in a collection? Are the results produced in such a way that they overwrite data?

You can also [pair an aggregation pipeline with a write operation that will change data in MongoDB](https://docs.mongodb.com/manual/tutorial/update-documents-with-aggregation-pipeline/).

allows you to create a pipeline that consists of one or more stages,
each of which performs a specific operation on your data.

You can think of the aggregation framework as similar to an automobile factory.
Automobile manufacturing requires the use of assembly stations organized
into assembly lines. Each station has specialized tools, such as
drills, welders, or other machinery. The factory transforms and
assembles the initial parts and materials into finished products.

The **aggregation pipeline** is the assembly line, **aggregation stages** are the assembly stations, and
**operator expressions** are the specialized tools.

Aggregation vs. Query Operators
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Using ``find`` operations, you can:

- select *which documents* to return
- select *which fields* to return
- sort the results

> Using ``aggregation`` operations, you can:

Author: accepted

- perform all ``find`` operations
- rename fields
- calculate fields
- summarize data
- group values

The aggregation framework uses a variety of :manual:`expression
operators </reference/operator/aggregation/>` that 

> allow you to group and modify data to produce a single result.

Author: my logic here was to compare it to the find/query operators, but the comparison that was here before made it sound like the find operations were weak, so i agree that this sentence doesnt do much on its own anymore.

Aggregation operations have some :manual:`limitations </core/aggregation-pipeline-limits/>`:

- Returned documents must not violate the :manual:`BSON-document size limit </reference/limits/#mongodb-limit-BSON-Document-Size>`
  of 16 megabytes.

- Pipeline stages have a memory limit of 100 megabytes by default. If necessary, you may exceed this limit by setting the ``allowDiskUse``
  property of ``AggregateOptions`` to ``true``. See the
  :node-api-4.0:`AggregateOptions API documentation <interfaces/aggregateoptions.html>`
  for more details.

.. important:: ``$graphLookup`` exception

   The :manual:`$graphLookup
   </reference/operator/aggregation/graphLookup/>` stage has a strict
   memory limit of 100 megabytes and will ignore ``allowDiskUse``.

Useful References
~~~~~~~~~~~~~~~~~

- :manual:`Aggregation pipeline </core/aggregation-pipeline/>`
- :manual:`Aggregation stages </meta/aggregation-quick-reference/#stages>`
- :manual:`Operator expressions </meta/aggregation-quick-reference/#operator-expressions>`

Runnable Examples
-----------------

> Create a new Node.js file called ``AggTour.js`` and connect to your

Author: I actually think there is no reason to even name the file, I simplified it

MongoDB deployment to follow this example.

.. tip::

   For information on connecting to your MongoDB deployment, see the :ref:`Connection Guide <connection>`.
   
Insert Sample Data
~~~~~~~~~~~~~~~~~~

The example uses a collection of documents corresponding
to information about 10 dining establishments. Insert the sample data

> into the ``restaurants`` collection of the ``aggregationTest`` database:

Author: Disagree on the last bullet point, I think writing out the names of the db and collection explain what's happening in those first lines of the code snippet

> into the ``restaurants`` collection of the ``aggregationTest`` database:

Author: accepted the rest of the suggestions

> into the ``restaurants`` collection of the ``aggregationTest`` database:

Reviewer: Won't push any further to have collection name and database name removed, but do those lines need to be included in the code snippet? I think we sometimes do this and sometimes don't, so totally up to you. 

What would be more helpful to readers (comprehensive content or focused content)? No clue what the answer to this is, and we spend a lot of time discussing this haha.

> .. code-block:: javascript

Author: accepted

>    const db = client.db("aggregationTest");

Author: accepted

   const coll = db.collection("restaurants");
   
   const docs = [
        { "name": "Sun Bakery Trattoria", "stars": 4, "categories": ["Pizza", "Pasta", "Italian"] },
        { "name": "Blue Bagels Grill", "stars": 3, "categories": ["Bagels", "Cookies", "Sandwiches"] },
        { "name": "Hot Fresh Bagels", "stars": 4, "categories": ["Bagels", "Sandwiches", "Coffee"] },
        { "name": "Cafe au Late", "stars": 4, "categories": ["Bakery", "Cafe", "Coffee", "Dessert"] },
        { "name": "Green Feast Pizzeria", "stars": 2, "categories": ["Pizza", "Italian"] },
        { "name": "Go Pasta", "stars": 0, "categories": ["Pasta", "Italian", "Cafeteria"] },
        { "name": "Liz's Coffee Bar", "stars": 5, "categories": ["Coffee", "Cafe", "Bakery"] },
        { "name": "Oak Steakhouse", "stars": 0, "categories": ["Steak", "Seafood"] },
        { "name": "Petit Cookie", "stars": 4, "categories": ["Bakery", "Cookies", "Cake", "Coffee"] },
        { "name": "Modern Eats", "stars": 3, "categories": ["Steak", "Salad", "American"] }

>     ];

Author: Reduced the number of entries to 5, left the names in

    // This option prevents additional documents from being inserted if one fails.
    const options = { ordered: true };

>     const result = await coll.insertMany(docs, options);

Author: accepted

Aggregation Example
~~~~~~~~~~~~~~~~~~~

To perform an aggregation, pass a list of aggregation stages to the 
``collection.aggregate()`` method.

> In the following example, the aggregation pipeline makes use of the

Reviewer: ```suggestion
In the following example, the aggregation pipeline uses the
```

I think "makes use" is idiomatic and may confuse non-native speakers. [See the style guide entry on writing for a global audience](https://docs.mongodb.com/meta/style-guide/terminology/terms-for-global-audience/#don-t-use-idioms-or-colloquialisms) for a full write up on this.

Author: accepted

following aggregation stages:

- A :manual:`$match </reference/operator/aggregation/match/>` stage to filter for documents whose
  ``categories`` array field contains the element ``Bakery``.

- A :manual:`$group </reference/operator/aggregation/group/>` stage to group the matching documents by the ``stars``
  field, accumulating a count of documents for each distinct value of ``stars``.

.. code-block:: javascript
   
   const pipeline = [
       {
           '$match': {
               'categories': 'Bakery'
           }
       }, {
           '$group': {
               '_id': '$stars',
               'count': {
                   '$sum': 1
               }
           }
       }
   ];

   const aggCursor = coll.aggregate(pipeline);
   await aggCursor.forEach(doc => console.log(doc));

> This operation should produce the following results:

Author: accepted, changed wording

===========
Aggregation
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use **aggregation operations** in
the MongoDB Node.js driver.

> Aggregation operations function to process data in your MongoDB

Author: modified to give a definition, but its hard because aggregation operations are really defined by what they do

collections and return computed results. MongoDB's aggregation framework
allows you to create a pipeline that consists of one or more stages,
each of which executes a specific operation on your data.

> You can think of the aggregation framework as similar to an automobile factory.

Author: car factory just doesn't sound like something i've ever heard

> You can think of the aggregation framework as similar to an automobile factory.

Author: i can change the other instances of automobile to car though

> You can think of the aggregation framework as similar to an automobile factory.

Reviewer: No worries! Sounds alright to my ear but I can see arguments for either. Totally cool if you want to keep automobile.

Automobile manufacturing requires the use of assembly lines that are

> comprised of assembly stations. Each station is equipped with specialized tools, such as

Author: tried to fix, wording might still be awkward

drills, welders, or other machinery. The factory transforms and

> assembles the initial raw parts and materials into finished products.

Author: accepted

The **aggregation pipeline** is the assembly line, **aggregation stages** are the assembly stations, and
**operator expressions** are the specialized tools.

> Aggregation vs. Find Operations

Author: accepted

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Using ``find`` operations, you can:

- select *which documents* to return
- select *which fields* to return
- sort the results

Using ``aggregation`` operations, you can:

- perform all ``find`` operations
- rename fields
- calculate fields
- summarize data
- group values

The aggregation framework includes a wide array of computational and

> descriptive operators that allow you to produce computed results,

Author: changed and simplified

whereas ``find`` operations provide a very limited number of data

> processing tools. 

Author: changed wording and removed comparison to find operations

> Note that aggregation operations have some :manual:`limitations </core/aggregation-pipeline-limits/>`:

Author: accepted

- Returned documents must not violate the :manual:`BSON-document size limit </reference/limits/#mongodb-limit-BSON-Document-Size>`
  of 16 megabytes.

- Pipeline stages have a memory limit of 100 megabytes by default. If necessary, you may exceed this limit by setting the ``allowDiskUse``
  property of ``AggregateOptions`` to ``true``. See the
  :node-api-4.0:`AggregateOptions API documentation <interfaces/aggregateoptions.html>`
  for more details.

.. important:: ``$graphLookup`` exception

   The :manual:`$graphLookup
   </reference/operator/aggregation/graphLookup/>` stage has a strict
   memory limit of 100 megabytes and will ignore ``allowDiskUse``.

Useful References
~~~~~~~~~~~~~~~~~

- :manual:`Aggregation pipeline </core/aggregation-pipeline/>`
- :manual:`Aggregation stages </meta/aggregation-quick-reference/#stages>`
- :manual:`Operator expressions </meta/aggregation-quick-reference/#operator-expressions>`

Runnable Examples
-----------------

Base Setup
~~~~~~~~~~

Create a new Node.js file called ``AggTour.js`` and include the following statements:

.. code-block:: javascript

   const MongoClient = require('mongodb').MongoClient;

>    const assert = require('assert');

Author: removed! not a Node user so i'm not always sure whats necessary in the code!

Connect to a MongoDB Deployment
+++++++++++++++++++++++++++++++

.. code-block:: javascript

   // Connect to your MongoDB deployment.

>    const uri = 'mongodb://localhost:27017';

Author: it doesn't have to be an atlas cluster, but for consistency i will change it to the template URI

   // Create or select an existing database.

>    const dbName = 'aggregationTest';

Author: accepted

   // Create a new MongoClient.
   const client = new MongoClient(uri);

   // Use connection method to connect to the Server.
   client.connect(function(err, client) {
        assert.equal(null, err);
        console.log("Connected correctly to server");
   });

.. tip::

   For information on connecting to your MongoDB deployment, see the :ref:`Connection Guide <connection>`.

Insert Sample Data
++++++++++++++++++

The examples in this guide use a collection of documents corresponding
to information about 10 dining establishments. Insert the sample data
into the ``restaurants`` collection of the ``aggregationTest`` database:

.. code-block:: javascript

>    // Select a database for use. 

Author: I agree- explanations are not needed. keeping the comment that states, "This option prevents additional documents from being inserted if one fails."

   const db = client.db(dbName);

   // Select or create a collection.

>    const coll = db.collection("restaurants");

Reviewer: Could these lines be moved to the earlier code block? Alternatively, I think it might make sense to remove the code about connecting to MongoDB and save that for another fundamentals page (Connection Guide Covers this in more depth).

Author: will restructure to link to the connection guide

        { "name": "Go Pasta", "contact": { "phone": "769-555-0152", "email": "GoPasta@example.com", "location": [-73.9446421, 40.7253944] }, "stars": 0, "categories": ["Pasta", "Italian", "Buffet", "Cafeteria"] },
        { "name": "Liz's Coffee Bar", "contact": { "phone": "644-555-0193", "email": "LizCoffeeBar@example.net", "location": [-74.0166091, 40.6284767] }, "stars": 5, "categories": ["Coffee", "Cafe", "Bakery", "Chocolates"] },
        { "name": "Oak Steakhouse", "contact": { "phone": "990-555-0165", "email": "OakSteakhouse@example.com", "location": [-73.9365108, 40.8497077] }, "stars": 0, "categories": ["Steak", "Seafood"] },
        { "name": "Petit Cookie", "contact": { "phone": "604-555-0149", "email": "PetitCookie@example.org", "location": [-73.8850023, 40.7494272] }, "stars": 4, "categories": ["Bakery", "Cookies", "Cake", "Coffee"] },
        { "name": "Modern Eats", "contact": { "phone": "229-555-0197", "email": "ModernEats@example.org", "location": [-73.9799932, 40.7660886] }, "stars": 3, "categories": ["Steak", "Salad", "American"] }

>     ];

Author: sure, I was following the example from the Java docs but the only relevant fields are name stars and cateories

    // This option prevents additional documents from being inserted if one fails.
    const options = { ordered: true };

    const result = await coll.insertMany(docs, options);

Basic Aggregation Example

> ~~~~~~~~~~~~~~~~~~~~~~~~~

Author: accepted

To perform an aggregation, pass a list of aggregation stages to the 
``collection.aggregate()`` method.

In the following example, the aggregation pipeline makes use of two

> aggregation stages:

Author: accepted

- A :manual:`$match </reference/operator/aggregation/match/>` stage to filter for documents whose
  ``categories`` array field contains the element ``Bakery``.

- A :manual:`$group </reference/operator/aggregation/group/>` stage to group the matching documents by the ``stars``
  field, accumulating a count of documents for each distinct value of ``stars``.

.. code-block:: javascript
   
   const pipeline = [
       {
           '$match': {
               'categories': 'Bakery'
           }
       }, {
           '$group': {
               '_id': '$stars',
               'count': {
                   '$sum': 1
               }
           }
       }
   ];

   const aggCursor = coll.aggregate(pipeline);
   await aggCursor.forEach(doc => console.log(doc));

This operation should produce the following results:

> .. code-block:: none

Author: accepted

   :copyable: false

   {"_id": 5, "count": 1}
   {"_id": 4, "count": 2}
   
For more information, see the :node-api-4.0:`aggregate() API documentation <classes/collection.html#aggregate>`.

Additional Aggregation Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
You can find another `aggregation framework example
<https://www.mongodb.com/blog/post/quick-start-nodejs--mongodb--how-to-analyze-data-using-the-aggregation-framework>`_

> in the Quick Start tutorials. 

Author: accepted
